### DISCUSSION TITLE
UI

### DISCUSSION THREAD

**Discussion by tronta** [phase1geo/Minder#384] [Ideas]
Hi *,

I'd like to start a discussion on how the UI of this exceptional great piece of SW could be improved, so it can be easier accessible for new user.
@phase1geo, please tell me, if this approach is ok for you? Hope this is not frightening. I tend to use holistic approaches. Not sure if eg the wiki might be a better place.

As a first step I would propose to have a common view on the current menu structure.
Discuss possible weaknesses and strength. How often functions are used and which of them context of a selected element.
Next, how it could be arranged, to minimize the weaknesses..

Tried to create an overview as (of course with a mindmap :smile: ), with it, and started with commenting my personal view on the menu.

![Overview current state](https://user-images.githubusercontent.com/5135577/124304993-c13ac980-db64-11eb-8556-307a0beb7a9d.jpg)

# Actions
A list of actions which can be triggered inside Minder
### Current
Is only used in context of a selected element. What to show if nothing is selected?
  - Task switcher
  - Fold/unfold
  - Color
  :arrow_forward: How is the default value set?
    * The default color of a node is the color of its parent node.  If the parent node is the root node of a tree, the color is selected on either a rotational basis of all available link colors or it will be set to the first available link color (depending on the map setting).
  - Note
  - Set Image
  - Copy/Paste, Link and Delete 
  :arrow_forward: could this live only in the context menu?
    * It could, but I thought it made sense to add some buttons for frequently used node functions in the Current sidebar.
### Style
Is only used in context of a selected element. What to show if nothing is selected?
[ ] How to set default values?
  - Branch Options
  :arrow_forward:  Do user use different settings in one mind map?
    - Style
    - Margin
  - Link Options
    - Line Dash
    - Line Width
    - Link Arrow
  - Node Options
    - Border Style
    - Border Width
    :arrow_forward: When would the user use this per node?
      * You could use it give a node some emphasis in the mind map.
    - Margin 
    :arrow_forward: There is already a margin as branch option, Which user really needs an other margin per node? 
      * Branch margin affects the distance between the parent node and the child node only.  Node margin affects the area around the entirety of the node.
    - Padding
    :arrow_forward: When would the user use this? When are the default settings not sufficient?
      * Padding adjustment can be used to make a node stand out in the mind map (it will affect the size of the node shape).  Using the padding globally can potentially make nodes easier to read as well, at the sacrifice of number of nodes displayed on the screen at one time.
    - Font
    - Width
    :arrow_forward: What is this for? When does the user use this?
      * This allows a selected node to be wider or narrower than the default width.
    - Enable Markup
    :arrow_forward: Is this really node specific? Could this be moved to the map preferences?
      * Allowing the user to adjust this on a node basis can improve application performance.
### Stickers
:heart: great overview of possible sticker, with fast access
  - search for stickers
  - favorite stickers
  - select stickers
### Map
  - Hide connections
  - Rotate main branch colors
  - Layout
  - Color Themes
  - Balance Nodes
  :arrow_forward: When would this be used?
    * For vertical or horizontal trees, node balancing will adjust the tree so that either top/bottom or left/right of the root node are roughly the same in size.  It can help make the tree more readable and printable.
  - Fold complete tasks
  - unfold all nodes

## Header
:broken_heart: A lot of icons on top level
### Undo
:question: I am not sure if this is needed, as most people know this key short cut
  * The tooltip for undo provides what action will be undone.  Without a UI element to attach the tooltip to, this information would be missing.
###  Redo
:question: I am not sure if this is needed, as most people know this key short cut
  * See above comment
### Focus Mode
:arrow_forward:  Great feature, but probably not used so often that it justifies to be on top level
  * I see focus mode as useful for doing presentations in addition to mind mapping itself.  For this reason, I chose to make the feature visible from the header bar, much like presentation software typically provides a button on the main interface that displays the slides in presentation mode.
### Zoom
- Zoom to Percent
- Zoom in
- Zoom out
- Zoom to fit
- Zoom to fit selected node
- Zoom to actual size
### Search
  - Search results
  - Search options
  - Search all tabs
    - Export
      - File Type
      - Export Options depending on Type
      - Print
### Gear
  - Preferences
    - Behavior
      - Create new mode from edit mode
      - Automatically make embedded URLs into links
      - Enable Markdown
      - Create connection title on creation
      - Select items on mouse over
    - Appearance
      - Default Theme
      - Enable animations
      - Text field font size
### Add/Edit
  - Test
  - Task
  - Image
  - Sticker
  - Group
  - Note
### Select 
- Root Node
 :arrow_forward: When does a user use this and not select it via mouse?
  * It's a quick way to get back to the root node of the tree.  If the tree is quite large, it can be cumbersome to pan the map in the attempt to find the root node.
- Next Sibling Node
- Previous Sibling Node
- Child Node
- Child Nodes
- Subtree
- Parent Node
- Linked Node
- Connection
### Sort Children
  - Alphabetically
  - Random
  :arrow_forward: When would a user use this function?
    * It was a user requested feature.  I think the function can sometimes be useful in the mind mapping process as it change things up from the original order to potentially spur new creative ideas.

# UI Top Level Proposal
![image](https://user-images.githubusercontent.com/5135577/124391382-ae550000-dcf0-11eb-8429-5e302289a122.png)
Not sure if the buttons `undo` and `redo` are really needed at top level, as most know and use people the short cuts.
`Map` Sidetab would move into document properties.
  * The Map sidebar tab controls the display attributes of the mind map.  The "document properties" menu that you list is mostly a list of functional things you can do with the mind map (i.e., save, save as, export, print, etc.).  In my mind, they are not related to each other.

## Open files
here I would take over Gnome's text editor design:
![image](https://user-images.githubusercontent.com/5135577/124395539-039b0c80-dd05-11eb-85c4-a71ffd66e5d0.png)

Where one can select the last opened files, plus open the full blown file selector.

## Zoom
The zoom options could look like this and make the menu smaller and maybe easier to read:
![image](https://user-images.githubusercontent.com/5135577/124416375-281fd480-dd57-11eb-8553-08fa8959ad00.png)
Maybe it would even make sense to reduce the dropdown list to a button....

:heart: I do like the addition of the plus/minus buttons to either side of the zoom level slider.  This allows us to remove the zoom in/out menu options and provides a UI element that we can attach a tooltip to which provides the keyboard shortcut for controlling this.  I think this is something that I am interesting in implementing in the elementary UI.
:arrow_forward: Displaying the current zoom level in the header bar seems like a waste of space in the header bar.  Perhaps this makes zoom features conform to the Gnome HIG, but it seems pretty senseless to me.

## Document Properties
![image](https://user-images.githubusercontent.com/5135577/124497355-03f5df00-ddbb-11eb-91b4-ced155878a32.png)

## Application Preferences
![image](https://user-images.githubusercontent.com/5135577/124493781-4ff25500-ddb6-11eb-8bf8-3bc0f145ffbc.png)

:arrow_forward: What are the three icons across the top of this popover?  I'm assuming the first icon will bring up the print dialog, but I'm not familiar with the other buttons displayed here.
:arrow_forward: I'm not a fan of placing the show/hide sidebar action in this menu.  It means that the user will need to first click the button and then click the sidebar menu option instead of a single click.  Showing and hiding the sidebar are things that the user will most likely want to do often when editing/reviewing the mind map.
:arrow_forward: Quite frankly, I don't like applications that create a top-level window containing functions that operate on the main window -- like "Find" and "Find and Replace".  It adds burdensome window management to an otherwise simple process.  It also creates separation of the function and document that don't need to be there.  I prefer applications that show/hide functionality within the main window based on user need, placed in a location that is convenient and makes sense.  Open/save dialogs and print dialogs are exceptions to this, however.  It is more important to use these system provided dialogs, and these windows don't interact with the document in the main window.

**Comment by tronta** [comment]
@phase1geo: I finished the first draft of the upper menus. What do you think so far?

**Reply by phase1geo** [reply]
I have provided input to questions above and added a few questions/comments of my own on other UI suggestions provided here.

**Comment by tronta** [comment]
Thanks for your comment. Maybe some additional words from my side:
I actually don't see contradicting design philosophies between Gnome and Elementary, Elementary is even referencing on the Gnome HIG site https://docs.elementary.io/hig/ for content they have not detailed yet. The problem is rather, that guidelines are never detailed enough (I know it's a lot of work, no offence). Or did I miss something?

Basically I tried to achieve the following things:
* [Avoid Feature Bloat](https://docs.elementary.io/hig/design-philosophy/concision): I'm sure, that there is a reason for each of the options. But I worked quite some time already with you software (thanks again for your great work), but imho, some of the options could be removed. Now telemetry would be great... but even without it, I could asking other users, how they are using the app.
* moving not that often used options from the top of the menus, further down. Eg header-bar buttons, to simplify the overall view.
* reuse common design patterns patters from other application, users are used to. Eg. search and replaced as two menu entries, which can be found almost everywhere like this.

[The key point, to draw the right assumptions,](https://docs.elementary.io/hig/design-philosophy/what-design-is-not) is to measure the design more objective.
I would be willing to make a questionnaire and ask other users about feedback. Or it could be even integrated as a feedback option inside the application (Link to Google Forms or similar !?)

As this is of course a lot of work, I would only do this, if you see benefits and would be committed to its outcome.
If your opinion is, that you want to design it for your personal needs and UI perception needs, this is totally fine as well. Do not misunderstand me. It's your time and application. I totally respect that.

**Reply by Nico7as** [reply]
Hi you both,

From a GNOME user standpoint, I really like Tronta's propositions. 

From an Elementary curious, I think the original design really fits into Elementary guidelines.

In an Ideal World, there would be enough developpers to develop both versions :-)

unfortunately, I'm no developer, I just wish I could have a more GNOME styled Minder!

Cheers to you both, keep up the good work!!
